pkg/controller/buildrun/runtime_image.go

```diff
@@ -0,0 +1,194 @@
+	// runtimeDockerfileTmpl Dockerfile template to be used with runtime-image, it uses Build attributes
+	// directly as template input.
+	runtimeDockerfileTmpl = `FROM {{ .Spec.Output.ImageURL }} as builder
+
+FROM {{ .Spec.Runtime.Base.ImageURL }}
+
+{{- range $k, $v := .Spec.Runtime.Env }}
+ENV {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $k, $v := .Spec.Runtime.Labels }}
+LABEL {{ $k }}="{{ $v }}"
+{{- end }}
+
+{{- range $dir := .Spec.Runtime.Directories }}
+{{- $parts := splitDirectories $dir }}
+COPY --from=builder "{{ index $parts 0 }}" "{{ index $parts 1 }}"
+{{- end }}
+
+{{- if .Spec.Runtime.WorkDir }}
+WORKDIR "{{ .Spec.Runtime.WorkDir }}"
+{{- end }}
+
+{{- if .Spec.Runtime.Entrypoint }}
+ENTRYPOINT [ {{ renderEntrypoint .Spec.Runtime.Entrypoint }} ]
+{{- end -}}
+`
+
+	// workspaceDir common workspace directory, where the source code is located.
+	workspaceDir = "/workspace/source"
+
+	// kanikoStable stable buildah image.
+	kanikoStable = "gcr.io/kaniko-project/executor:v0.23.0"
+
+	// runtimeDockerfile runtime Dockerfile file name.
+	runtimeDockerfile = "Dockerfile.runtime"
+)
+
+// rootUserID root's UID
+var rootUserID = int64(0)
+
+// runtimeDockerfilePath path to runtime Dockerfile on workspace directory.
+var runtimeDockerfilePath = path.Join(workspaceDir, runtimeDockerfile)
+
+// splitDirectories split informed directory by colon, returning a slice with parts. When colon is
+// not present, return the informed directory twice. This method always returns a slice with two
+// entries.
+func splitDirectories(dir string) []string {
+	parts := strings.Split(dir, ":")
+	if len(parts) == 2 {
+		return parts
+	}
+	return []string{dir, dir}
+}
+
+// renderEntrypoint will take a slice of strings and render the notation expected on ENTRYPOINT.
+func renderEntrypoint(e []string) string {
+	entrypoint := []string{}
+	for _, cmd := range e {
+		entrypoint = append(entrypoint, strconv.Quote(cmd))
+	}
+	return strings.Join(entrypoint, ", ")
+}
+
+// renderRuntimeDockerfile render runtime Dockerfile using build instance and pre-defined template.
+func renderRuntimeDockerfile(b *buildv1alpha1.Build) (*bytes.Buffer, error) {
+	tmpl, err := template.New(runtimeDockerfile).
+		Funcs(template.FuncMap{
+			"splitDirectories": splitDirectories,
+			"renderEntrypoint": renderEntrypoint,
+		}).
+		Parse(runtimeDockerfileTmpl)
+	if err != nil {
+		return nil, err
+	}
+
+	dockerfile := new(bytes.Buffer)
+	if err = tmpl.Execute(dockerfile, b); err != nil {
+		return nil, err
+	}
+	return dockerfile, nil
+}
+
+// runtimeDockerfileStep trigger the rendering of Dockerfile.runtime, and use this input as a
+// build-step to create a new file.
+func runtimeDockerfileStep(b *buildv1alpha1.Build) (*buildv1alpha1.BuildStep, error) {
+	dockerfile, err := renderRuntimeDockerfile(b)
+	if err != nil {
+		return nil, err
+	}
+
+	container := v1.Container{
+		Name:  "runtime-dockerfile",
+		Image: b.Spec.BuilderImage.ImageURL,
+		SecurityContext: &v1.SecurityContext{
```

Comment:
Admittedly somewhat of a tangential point, but the current approach of embedding and exposure of the entire corev1.Container type into the `BuildStep` will just continue to proliferate @sbose78 @otaviof @adambkaplan as we add feature to this repo, and just potentially widens our escalation window.

I think the sooner we abstract out, in as fine grained and focused a manner as possible, the bits of corev1.Container we need, and wherever possible have the controller just set the fields for the user vs. providing a choice, the better.

Don't see a feature along these line via https://github.com/redhat-developer/build/issues?q=is%3Aissue+is%3Aopen+label%3Aenhancement

Is everyone amenable to my opening one?  Is there enough agreement that we should get a Jira story on our backlog to accompany the github enhancement issue?

thanks

Reply:
@gabemontero pls refer to https://github.com/redhat-developer/build/issues/269#issuecomment-644167359 and pls open an issue if needed to tackle that concern. I dont think this PR is the right place for that.

Reply:
mostly agreed as I tried to clarify in my comment @qu1queee 

I do think citing how the proliferation is occurring is relevant.

And the Jira vs. github issues has bearing priority wise for the team here at RH.  I'm actually more interested in this becoming a Jira.

But assuming @sbose78 @adambkaplan don't miss this thread in their github notifications, I'm fine with you marking this resolved if you like.  I'll follow up with them separately if they do not respond here.

Reply:
I agree, we should support a minimal container type. We don't necessarily need a full-fledged EP to begin with I feel. 

I am OK if we make a PR to have a minimal container type with only the things that have been in use so far :)

Reply:
Of course, let's ensure we have a separate issue & jira tracked for this.